### PR TITLE
Fix issue with incorrectly inferred Scala version

### LINKED ISSUE #27226: Scala plugin infers an incorrect Scala version
### Current Behavior

The `scala` plugin infers the Scala version incorrectly if the project uses a Scala library JAR with a nonstandard name, causing the build to fail. Because of this, the `scala` plugin can't be used with Ivy repositories where the assumption about artifact file names doesn't hold.

Example Ivy artifact path of the Scala library JAR:
```
org.scala-lang/scala-library/2.13.12/jars/scala-library-2.x.jar
```
Resulting error message:
```
* What went wrong:
Execution failed for task ':compileScala'.
> Could not resolve all files for configuration ':detachedConfiguration1'.
   > Could not find org.scala-lang:scala-compiler:2.x.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/scala-lang/scala-compiler/2.x/scala-compiler-2.x.pom
     Required by:
         project :
   > Could not find org.scala-sbt:compiler-bridge_2.x:1.9.3.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/scala-sbt/compiler-bridge_2.x/1.9.3/compiler-bridge_2.x-1.9.3.pom
     Required by:
         project :
```

### Expected Behavior

I see two options. The first would be to give the users control over the version, and would make it much easier to solve any issues in customized environments. However if we had to choose, we would prefer the second solution as that would work automatically out of the box. (It would probably be best to implement both, if possible.)
1. There should be an API to specify it explicitly (see also: #18435).
2. The `scala` plugin should infer the Scala version correctly if there is a Scala Library JAR on the classpath, regardless of the name of that JAR file. This is possible by looking at the contents of the JAR:
   * The `2.x` JARs (e.g. [`scala-library-2.13.12.jar`](https://mvnrepository.com/artifact/org.scala-lang/scala-library/2.13.12)) contain a `library.properties` ZIP entry, that has a `version.number=2.13.12` line stating the version.
   * The `3.x` JARs (e.g. [`scala3-library_3-3.3.1.jar`](https://mvnrepository.com/artifact/org.scala-lang/scala3-library_3/3.3.1)) contain a `META-INF\MANIFEST.MF` ZIP entry, that has a `Implementation-Version: 3.3.1` line stating the version.
   * If reading the known-good version from the JAR fails for whatever reason, the implementation can still fall back to parsing the file name.

### Context (optional)

The reproduction (`kotlin-dsl-reproduction` branch) uses a local Ivy repo to be self-contained, but it would fail in a very similar way if the same Ivy repo was on an HTTP server. In that case, the path and name of the Scala library JAR is whatever Gradle generates for its local cache (`.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-library/2.13.12/3c51da898af4ebfac356fcde11388afb2ae8ee82/scala-library-2.x-2.13.12.jar`), but unfortunately that also can't be parsed correctly: it results in version `2.x-2.13.12` being inferred. You can try this yourself by changing the Ivy repo's URL to `https://raw.githubusercontent.com/pantherdd/gradle-issue-scala-version-inference/kotlin-dsl-reproduction/ivy-repo`.

I also found a dirty, hacky workaround (`kotlin-dsl-workaround` branch), that allows us to proceed for the time being, but unfortunately it doesn't even solve all issues. For example, IntelliJ's Gradle integration also calls `ScalaRuntimeHelper.getScalaVersion` with an input we have no control over, which results in `Warning: Cannot determine Scala compiler version for module reproducer-project.main`.

### Steps to Reproduce

1. https://github.com/pantherdd/gradle-issue-scala-version-inference/tree/kotlin-dsl – Basic Scala project that works
2. https://github.com/pantherdd/gradle-issue-scala-version-inference/tree/kotlin-dsl-reproduction – Reproduction of the issue
3. https://github.com/pantherdd/gradle-issue-scala-version-inference/tree/kotlin-dsl-workaround – Workaround

### Gradle version

8.4

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by pantherdd**
This pull request fixes the following issue:
* https://github.com/gradle/gradle/issues/27226

### Context

This is a follow-up to #28198 , but [as requested there](https://github.com/gradle/gradle/pull/28198#issuecomment-2077337116), I'm only trying to fix #27226 on its own first, without adding new public APIs or creating named configurations.

We can later continue and tackle the other issues related to Scala classpaths in separate PRs.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [x] Provide integration tests (under `scala/src/integTest`) to verify changes from a user perspective.
- [x] Provide unit tests (under `scala/src/test`) to verify logic.
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew scala:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/jvm/scala/src/main/java/org/gradle/api/tasks/scala/ScalaTask.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.gradle.api.Task;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.tasks.Classpath;
+import org.gradle.internal.instrumentation.api.annotations.ToBeReplacedByLazyProperty;
+
+/**
+ * Common interface for Scala tasks.
+ *
+ * @since 8.10
```
Comment (pantherdd) [inline platforms/jvm/scala/src/main/java/org/gradle/api/tasks/scala/ScalaTask.java:27]: Let me know whether these changes still have a chance to make it into `8.10` – if not, I'll have to update this.

**Comment by pantherdd** [comment]
@big-guy I believe I was able to fix https://github.com/gradle/gradle/issues/27226 in this PR in accordance with your comments ([here](https://github.com/gradle/gradle/issues/27226#issuecomment-1853049118) and [here](https://github.com/gradle/gradle/pull/28198#issuecomment-2077337116)).
* Public APIs are not modified.
* The implementation still uses detached configurations (not moving to named ones, we can discuss options [later](https://github.com/gradle/gradle/issues/27309)).
* Version inference gets the information from the property files inside the Scala JARs, where possible.
* While I understand that inferring the version is not a strategic direction, I don't have the time to do a major refactor of the Scala plugin, where all of that would go away and a DSL would replace it. I'm just trying to fix bugs, while keeping the current behavior intact, without breaking any existing builds.

Can you check and let me know whether I have covered all the blocking issues that got #28198 declined?

Thanks!

**Comment by ljacomet** [comment]
Thank you for your contribution!

This PR has a valid DCO and tests. The relevant team for this area will confirm the design of the implementation choices.

**Comment by pantherdd** [comment]
@big-guy Would I need to fix the conflict before you can review the changes, or is it okay to do it afterwards?

**Comment by pantherdd** [comment]
@ljacomet Would I need to fix the conflict before someone can review the changes, or is it okay to do it afterwards?

**Comment by pantherdd** [comment]
What is the next step?

**Comment by big-guy** [comment]
Hi @pantherdd, don't worry about the merge conflict.  accepted-public-api-changes.json changes all the time and we can deal with it later. 

This is on my list to review and I've taken an initial look and don't see any major blockers. I mostly need to remind myself again how this all worked.

**Comment by pantherdd** [comment]
@big-guy can you share any estimate when you expect to have time for this?

**Comment by lmlynik** [comment]
Hi @pantherdd , do you think this PR would possibly fix the issue we encountered in https://github.com/gradle/gradle/issues/30419?
I believe issue is related to missing Scala 3.5.0 missing in bootclasspath and this PR vaguelly seems to refer to that

**Comment by pantherdd** [comment]
@lmlynik Possibly, it depends on what exactly is going wrong in your case. You could try debugging into `ScalaRuntime.inferScalaClasspath` and see what value gets assigned to `scalaVersion` and why.

**Comment by pantherdd** [comment]
@big-guy can you share any estimate when you expect you'll have time for this?

**Comment by pantherdd** [comment]
@big-guy @ljacomet @ghale @reinsch82 @jvandort
This PR is in the "ready for review" state since July 16th. 2 months have passed but there does not seem to be any progress.
What am I doing wrong here? Please advise.

**Comment by pantherdd** [comment]
Can you share an ETA please?

**Comment by jvandort** [comment]
We do not want to double-down on scanning the classpath to determine the scala version. 

Please see this PR which introduces an API for the user to declare the scala version. The plugin adds the scala dependencies for the user, so that it does not need to infer the scala version at all: https://github.com/gradle/gradle/pull/31260

**Pull request closed without merging**